### PR TITLE
feat!: add Gate audio effect and unify the dynamics node scaffolding

### PR DESCRIPTION
**Pull request by yuto-trd**
## Summary

Adds a **Gate** (noise gate) audio effect, completing the dynamics-processing set alongside the already-merged Compressor (#1724) and Limiter (#1725), then unifies the scaffolding the dynamics nodes had been duplicating:

- **`GateEffect`** (new, `Beutl.Engine.Audio.Effects`) — animatable `Threshold` / `Attack` / `Hold` / `Release` / `Range` properties, following the `CompressorEffect` surface pattern (`ScanProperties`, `[Range]`/`[Display]` attributes, `CreateNode` via `AddNode` + `Connect`).
- **`GateNode`** (new, `Beutl.Engine.Audio.Graph.Nodes`) — per-sample DSP: static/animated paths sharing one gate-gain helper, linked peak detection across all channels, hold-timer latching, one-pole dB-domain attack/release, `Range = 0` as an exact identity, and a below-threshold lead-in seeded at the Range floor.
- **`GateParameters`** (internal) — shared ranges/defaults, same convention as `CompressorParameters`/`LimiterParameters`.
- **`DynamicsNode`** (new, public abstract) — the scaffolding `CompressorNode` and `GateNode` each carried a copy of: chunk validation, session/seek state resets, the per-channel buffer view cache, parameter sanitization and the warn-once diagnostics. Each node now supplies only its parameter set and gain math. Net −82 lines of production code.
- **`AudioProcessContext.ContinuesFrom`** (new) — single home for the chunk-contiguity check, replacing three copies of the same one-tick tolerance and fixing two nodes that did not have it (see below).
- `LibraryRegistrar` registration + EN/JA display strings, so Gate appears in the audio-effect add menu (parameter UI is auto-generated by `AudioEffectEditor`).

`LimiterNode` deliberately does **not** derive from `DynamicsNode`: its lookahead delay lines and per-parameter error-severity diagnostics (`ref bool` latches, per-parameter fallbacks) are a different contract, not a variation of this one. It does share `ContinuesFrom`.

### Fixes to existing nodes

Pulled in because the shared base and the shared contiguity check made the divergences visible:

- **`DelayNode` / `EqualizerNode`** compared chunk timestamps exactly while Gate, Compressor and Limiter tolerate the one-tick difference independent `TimeSpan` rounding can put between adjacent sample boundaries. In a chain mixing them, one scrub dropped the delay lines and the biquad state while the dynamics nodes kept theirs — an audible click from a rounding artifact rather than a seek.
- **`CompressorNode`** gains four behaviours from the shared base: it rejects a sample-rate mismatch, tolerates one-tick boundary rounding, falls back to the declared constant when a property's default is itself non-finite, and no longer lets an empty chunk mask a later discontinuity.
- **`GateNode`**: an empty chunk (spans time, carries no samples) advanced `_lastTimeRangeEnd`, so the chunk after a hole looked contiguous with the chunk before it and inherited stale gate state.

### Known behaviour, deliberately unchanged

Peak detection is unsmoothed, so `Hold` is what bridges a waveform's zero crossings. At `Hold = 0` with the fastest `Release`, a low-frequency tone that never leaves the open region still pumps (50 Hz: minimum gain 0.069; the defaults measure 1.000). Real gates behave the same way at those settings, and smoothing the detector would need a time constant that duplicates `Hold`'s role — so the behaviour stays and is pinned by a characterization test plus a comment on `MinHoldMs`.

### Board-item scope note

The board item's 現状把握 predates #1724/#1725: Compressor and Limiter are already on `main` with full tests, so only Gate remained in the dynamics family. **NoiseSuppression (SpectralGate)** is a distinct FFT/STFT architecture that the item body itself splits into a separate phase; it is intentionally not part of this PR (no TODO markers left in code — the phase boundary is documented here and in the commit body). Future naming is collision-free (`SpectralGate*` vs `Gate*`).

## Breaking changes

`Beutl.Engine`:

- `CompressorNode` and `GateNode` now derive from the new public abstract `DynamicsNode` rather than directly from `AudioNode`. Treating them as `AudioNode` is unaffected; `Reset`, the channel-view cache and the parameter sanitizers now live on the base.
- `CompressorNode` throws `InvalidOperationException` when the upstream buffer's sample rate does not match the context's. It previously processed the buffer and mislabelled its samples. `GateNode` and `LimiterNode` already rejected this.

Plugin authors gain `DynamicsNode` as an extension point: deriving from it provides chunk validation, seek/session resets, the per-channel view cache and the warn-once parameter diagnostics without reimplementing them.

## Tests

- `GateNodeTests` + `GateEffectTests` — 51 tests: silence input, threshold open/close, attack time constant, hold latching down to a single sample, Range floor and exact `Range = 0` identity, chunk-boundary continuity, seek/sample-rate resets, mono and 4-channel linked gating, static-vs-animated equivalence (stereo and surround), non-finite input and parameter sanitization, diagnostic latch re-arm semantics, and the zero-hold characterization above.
- `CompressorNodeTests` +5, `AudioProcessContextTests` +6, `DelayNodeTests` +2, and a new `EqualizerNodeTests` fixture (+3) covering the behaviours listed under "Fixes to existing nodes". `EqualizerNode` had no test fixture before this PR.
- Every one-tick test was verified to actually fail when the tolerance is set to 0 (11 tests across all five stateful nodes), and the surround and empty-chunk tests were verified against deliberately broken code. Two of them were rewritten after that check showed they passed regardless: the surround static/animated comparison needed per-channel levels where only one channel crosses the threshold, and the Delay test needed a time-independent input because `RampInputNode` derives its samples from the chunk start time (at the fixture's 100 Hz, one tick of offset shifts the ramp by a whole sample).
- `Engine.Audio`: 267 passed / 0 failed. Full `Beutl.UnitTests`: 3928 passed / 0 failed.
- Local gates: `dotnet build Beutl.slnx` 0 errors; `dotnet format --verify-no-changes` clean; UTF-8 BOM on all new `.cs` files.

## Review

- Pre-PR design review: **approve** (pass 1, zero rework findings) — pattern-consistent with Compressor/Limiter, orthogonal property vocabulary.
- That review noted, as out of scope, that the dynamics nodes shared duplicated scaffolding that could someday become a joint `DynamicsNode` refactor. A follow-up review of this branch raised the same duplication plus five smaller findings (untested >2-channel paths, test-only counters with no test, unreachable defensive code, the empty-chunk ordering, and the unsmoothed detector); all six are addressed here.

## Manual verification

1. Real-time playback: gate an intermittent-speech clip, confirm no clicks at open/close transitions across buffer boundaries.
2. Export parity with preview (same `GateNode` drives both).
3. Scrub a chain that mixes Gate/Compressor with Delay/Equalizer and confirm no click at chunk boundaries (the contiguity fix above).
4. Undo/redo of parameter edits (CoreProperty-based).
5. "Gate" appears in the audio-effect add menu with correct EN/JA labels.

Refs: Project #9 board item「オーディオエフェクト拡充 (Compressor / Limiter / NoiseSuppression)」

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a new audio **Gate** effect with animatable **Threshold, Attack, Hold, Release,** and **Range** controls.
  * Added localized UI labels and descriptions for gate parameters (including Japanese) and included the effect in the audio effect library.
* **Bug Fixes**
  * Improved gate stability and continuity across renders, chunk boundaries, and sample-rate changes.
  * Enhanced handling for digital silence, non-finite values, and exact **Range = 0** passthrough.
* **Tests**
  * Added unit tests covering effect/node wiring, default values, and both static and animated processing paths.

### CODE REVIEW

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+
+public sealed class GateNode : AudioNode
+{
+    private static readonly ILogger s_logger = Log.CreateLogger<GateNode>();
+
+    private const float MinDb = -100f;
+
+    // Gate gain in dB, smoothed toward 0 (open) or the Range floor (closed). Starts fully closed so a
+    // silent lead-in stays quiet until the signal crosses the threshold.
+    private float _gateGainDb = MinDb;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs:15]: Confirmed and fixed in de710f38b. With Range=0 the one-pole follower started at -100 dB and only asymptotically approached 0 dB, producing a fade-in and perpetual sub-unity gain instead of the documented "gating disabled". NextGain now snaps the follower open and returns unity when Range >= 0, so output equals input sample-for-sample. Regression test: Process_RangeZero_IsExactIdentity (exact equality, red on the previous commit).

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+            {
+                Span<float> in0 = inputChannels[0].Span;
+                Span<float> out0 = outputChannels[0].Span;
+                Span<float> in1 = channels == 2 ? inputChannels[1].Span : default;
+                Span<float> out1 = channels == 2 ? outputChannels[1].Span : default;
+
+                for (int i = 0; i < sampleCount; i++)
+                {
+                    float s0 = in0[i];
+                    float peak = MathF.Abs(s0);
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs]: Confirmed and fixed in de710f38b. The mono/stereo fast path seeded peak = MathF.Abs(s0), so a NaN in channel 0 poisoned the linked-stereo peak (a1 > NaN is false). Both fast paths now seed 0 and compare, matching the already-safe >2-channel path. Regression test: Process_OneChannelNaN_GatesFromValidChannel. Note: CompressorNode on main shares this fast-path shape; tracked separately as it is outside this PR's scope.

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+        }
+        return 0f;
+    }
+
+    // Advances the gate one sample and returns the linear gain. Shared by ProcessStatic and
+    // ProcessAnimated so the gate/hold math cannot drift between the paths.
+    private float NextGain(float peak, float attackCoeff, float releaseCoeff, in EffectiveParameters p, int holdSamples)
+    {
+        float inputDb = peak > 0f ? 20f * MathF.Log10(peak) : MinDb;
+        bool aboveThreshold = inputDb >= p.Threshold;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs]: Confirmed and fixed in de710f38b. Digital silence collapsed to inputDb = -100 which compared equal to the -100 dB minimum threshold and opened the gate. Opening now additionally requires peak > 0, so true digital silence stays gated at any threshold. Regression test: Process_MinThreshold_DigitalSilenceDoesNotOpenGate.

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+    /// </remarks>
+    public void Reset()
+    {
+        ResetGate();
+        ResetDiagnostics();
+    }
+
+    private void ResetGate()
+    {
+        _gateGainDb = MinDb;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs:299]: Fixed in f4bb29344. ResetGate seeded the follower at -100 dB, so a below-threshold lead-in with Range above MinDb (e.g. -20 dB) faded in from near-silence up toward the floor instead of resting at it. NextGain now seeds the follower at the effective Range floor on the first sample after a reset, so a closed lead-in sits at the user's Range immediately; an above-threshold onset still ramps open from that floor via Attack, and the Range >= 0 disabled path is unaffected. Regression test: Process_BelowThresholdLeadIn_StartsAtRangeFloorNotFullMute (first-sample gain equals the Range floor; red at -99.98 dB on the previous commit).

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+        if (_lastSampleRate != context.SampleRate)
+        {
+            Reset();
+            _lastSampleRate = context.SampleRate;
+        }
+
+        // Reset the gate on the first call or whenever this chunk does not continue from the previous
+        // one; the node is cached across Compose() calls, so stale gate state would otherwise bleed in
+        // after a seek/restart. Only DSP state resets here, not the diagnostic latches.
+        if (!_lastTimeRangeEnd.HasValue || _lastTimeRangeEnd.Value != context.TimeRange.Start)
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs]: Fixed in f4bb29344. The discontinuity check compared _lastTimeRangeEnd against TimeRange.Start with exact equality, so at 44.1 kHz a chunk whose start rounds one tick off previousEnd could be misread as a seek and reset the gate at regular buffer boundaries. GateNode now applies the same one-tick tolerance as LimiterNode (TimestampQuantizationToleranceTicks = 1 via IsTimestampContiguous); a two-tick-or-larger gap still resets. Regression test: Process_OneTickBoundaryRounding_DoesNotResetGate (a one-tick-early adjacent chunk continues the warmed-open gate; red on the previous commit where continuing == fresh).

File: tests/Beutl.UnitTests/Engine/Audio/GateNodeTests.cs [hunk 6]
```diff
@@ -0,0 +1,787 @@
...
+    {
+        // Adjacent sample-boundary chunks can differ by one tick from independent TimeSpan rounding. A
+        // one-tick gap must NOT be treated as a seek: the warmed-open gate must continue (its first
+        // sample stays loud) rather than reset to closed like a fresh gate. Buggy exact-equality would
+        // reset here, dropping the continuing gate's first sample to the fresh gate's value.
+        const int chunkSamples = SampleRate / 10;
+        var chunkDuration = TimeSpan.FromSeconds(chunkSamples / (double)SampleRate);
+        var ctx1 = CreateContext(TimeSpan.Zero, chunkDuration);
+        // Second chunk starts one tick before the exact previous end — within the rounding tolerance.
+        var ctx2 = CreateContext(chunkDuration - TimeSpan.FromTicks(1), chunkDuration);
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply tests/Beutl.UnitTests/Engine/Audio/GateNodeTests.cs]: Done in 5be4d8ad6. Parametrized the test with [TestCase(-1L)]/[TestCase(1L)] so it covers both a one-tick overlap (chunkDuration - 1 tick) and a one-tick gap (chunkDuration + 1 tick), matching the symmetric tolerance in IsTimestampContiguous (difference is >= -1 and <= 1). Both cases pass; an asymmetric implementation that reset on either side would now fail.

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+        // collapse to MinDb and compare equal (-100 >= -100). Require real signal energy to open.
+        float inputDb = peak > 0f ? 20f * MathF.Log10(peak) : MinDb;
+        bool aboveThreshold = peak > 0f && inputDb >= p.Threshold;
+        if (aboveThreshold)
+        {
+            _holdCounter = holdSamples;
+        }
+        else if (_holdCounter > 0)
+        {
+            _holdCounter--;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs]: Fixed in 7ce866d48. Confirmed: _holdCounter was decremented before the target-gain check read it, so a hold of N samples latched the gate open for only N-1 — and a hold converting to a single sample released immediately. NextGain now captures heldOpen = _holdCounter > 0 for the current sample and only then spends the counter, so every configured hold sample is preserved. Regression test: Process_Hold_KeepsGateOpenForEveryConfiguredSample (one-sample hold; the first below-threshold sample must stay at unity — red at 0.867 on the previous commit).

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+    private AudioBuffer ProcessStatic(AudioBuffer input, AudioProcessContext context)
+    {
+        var output = new AudioBuffer(input.SampleRate, input.ChannelCount, input.SampleCount);
+        try
+        {
+            EffectiveParameters p = ReadStaticParameters();
+
+            float attackCoeff = ComputeCoeff(p.Attack, context.SampleRate);
+            float releaseCoeff = ComputeCoeff(p.Release, context.SampleRate);
+            int holdSamples = HoldSamples(p.Hold, context.SampleRate);
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs:50]: Fixed in 7ce866d48. GateNode now validates input.SampleRate == context.SampleRate and throws InvalidOperationException otherwise, matching LimiterNode.cs:71-73. The coefficients and hold count are derived from context.SampleRate while the output buffer carries input.SampleRate, so the mismatch was silently mislabelling samples. Regression test: Process_InputSampleRateMismatch_Throws (a 44.1 kHz buffer under a 48 kHz context; red on the previous commit where no exception was thrown).

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+    private float SafeParameter(float value, float fallback, string paramName)
+    {
+        if (float.IsFinite(value)) return value;
+        if (_loggedNonFiniteParameters.Add(paramName))
+        {
+            s_logger.LogWarning(
+                "Gate parameter '{Param}' produced a non-finite value; falling back to {Fallback}. Further occurrences for this parameter will be suppressed.",
+                paramName, fallback);
+        }
+        return fallback;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs]: Fixed in 7ce866d48. SafeParameter now takes the declared GateParameters constant as a last-resort default and uses it when the property's DefaultValue is itself non-finite, so Math.Clamp is never handed a NaN it cannot sanitize. Regression test: Process_NonFinitePropertyDefault_FallsBackToDeclaredConstant (all five properties declared with float.NaN as both CurrentValue and DefaultValue; an above-threshold signal must still pass at unity — red at -100.9 dB on the previous commit, i.e. pinned to the recovery sentinel exactly as described).

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+
+    public required IProperty<float> Release { get; init; }
+
+    public required IProperty<float> Range { get; init; }
+
+    public override AudioBuffer Process(AudioProcessContext context)
+    {
+        if (Inputs.Count != 1)
+            throw new InvalidOperationException(
+                $"Gate node requires exactly one input but got {Inputs.Count}.");
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs]: Noted, but out of scope for this PR. Side-chain gating is a separate feature rather than a property of the noise gate being added here: exposing a detector input or a detector strategy changes GateNode's public input contract, and the same question applies equally to CompressorNode and LimiterNode (both are sealed, self-detecting, single-input today). Designing that surface for one node in isolation would leave the dynamics set inconsistent, so it belongs in a dedicated change that covers all three. This PR keeps the single-input, self-detecting contract that matches its siblings; no code change here.

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+                Span<float> in1 = channels == 2 ? inputChannels[1].Span : default;
+                Span<float> out1 = channels == 2 ? outputChannels[1].Span : default;
+
+                for (int i = 0; i < sampleCount; i++)
+                {
+                    float s0 = in0[i];
+                    // Comparisons against NaN stay false, preserving a finite channel's peak.
+                    float peak = 0f;
+                    float a0 = MathF.Abs(s0);
+                    if (a0 > peak) peak = a0;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs]: Fixed in 5ff715fd2. Peak detection now skips non-finite samples in both processing paths (`float.IsFinite(a) && a > peak`), so a corrupt channel can no longer drive the linked peak — `src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs:69` and the `>2`-channel fallback at `:96`, plus the same two sites in `ProcessAnimated`.

Covered by `Process_OneChannelInfinity_DoesNotOpenGateForValidChannels` (static and animated paths, single- and multi-sample corruption). Verified as a regression test: it fails against the pre-fix code on all three cases and passes after.

File: src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+                $"{DiagnosticName} node requires exactly one input but got {Inputs.Count}.");
+
+        // Every path emits a fresh buffer (no pass-through), so dispose the consumed input.
+        using var input = Inputs[0].Process(context);
+
+        // Coefficients come from context.SampleRate while the output carries input.SampleRate, so a
+        // mismatch would mislabel samples; this node does not resample.
+        if (input.SampleRate != context.SampleRate)
+            throw new InvalidOperationException(
+                $"{DiagnosticName} node: sample rate mismatch. context={context.SampleRate}, input={input.SampleRate}.");
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs]: Already addressed — the marker is on the commit that introduces the change, not the aggregate: `88abda60a` is `refactor!: extract shared DynamicsNode scaffolding` and carries a `BREAKING CHANGE:` footer naming both the `DynamicsNode` base-type change and this `InvalidOperationException` on a sample-rate mismatch.

The PR metadata was stale when this review ran and has since been updated: the title is now `feat!: add Gate audio effect and unify the dynamics node scaffolding`, the "non-breaking `feat:`" sentence is gone, and the body has a dedicated "Breaking changes" section describing the migration for downstream `Beutl.Engine` consumers.

File: src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+            ResetDspState();
+        }
+        _lastTimeRangeEnd = context.TimeRange.Start + context.TimeRange.Duration;
+
+        return HasAnimatedParameters ? ProcessAnimated(input, context) : ProcessStatic(input, context);
+    }
+
+    protected abstract AudioBuffer ProcessStatic(AudioBuffer input, AudioProcessContext context);
+
+    protected abstract AudioBuffer ProcessAnimated(AudioBuffer input, AudioProcessContext context);
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs:109]: Agreed in principle, but deferred out of this PR by an explicit scope decision — tracked as a Backlog draft on the AI Review board (Project #9): **"refactor(engine): let DynamicsNode own the per-sample processing loop"**.

Reasoning for not doing it here: hoisting the loops means the base owns the `stackalloc` scratch for N animated parameters and hands slices back to the derived node, since the two parameter sets differ in size and meaning (Gate: Threshold/Attack/Hold/Release/Range; Compressor: Threshold/Ratio/Attack/Release/Knee/MakeupGain). That trades duplicated-but-obvious DSP loops for index bookkeeping in hot audio code, and it reverses a design decision this PR made deliberately — so it deserves its own change rather than being appended here.

The board item records the sketch to evaluate (a generic `DynamicsNode<TParameters>` with `NextGain` as the only required hook), the open questions (where the coefficient cache lives; whether the detector needs a hook of its own, given Gate now skips non-finite samples in the peak scan while Compressor does not; whether `LimiterNode` can eventually join), and the definition of done.

File: src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+                    float s1 = 0f;
+                    if (channels == 2)
+                    {
+                        s1 = in1[i];
+                        float a1 = MathF.Abs(s1);
+                        if (a1 > peak) peak = a1;
+                    }
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs:72]: Correct, and thank you — the guard was missing exactly there. Fixed in 1c09b6497: `src/Beutl.Engine/Audio/Graph/Nodes/GateNode.cs:69` and `:75` now match the other three sites.

Root cause on my side was the test, not just the code: the matrix I added with the first fix only exercised 4-channel buffers, so it never entered the `channels <= 2` fast path and reported green against the unguarded code. `Process_OneChannelInfinity_DoesNotOpenGateForValidChannels` is now parameterised over channel layout as well — `(2,1,false)`, `(2,4,false)`, `(2,4,true)`, `(4,4,false)`, `(4,4,true)` — and the two stereo static cases fail against the pre-fix code, which is what was missing before.

File: src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+    protected abstract bool HasAnimatedParameters { get; }
+
+    public override AudioBuffer Process(AudioProcessContext context)
+    {
+        if (Inputs.Count != 1)
+            throw new InvalidOperationException(
+                $"{DiagnosticName} node requires exactly one input but got {Inputs.Count}.");
+
+        // Every path emits a fresh buffer (no pass-through), so dispose the consumed input.
+        using var input = Inputs[0].Process(context);
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs]: Valid — fixed in 1c09b6497 by preserving ownership rather than forbidding pass-through, since `AudioNodeBufferDisposalTests` already pins pass-through as the repo-wide contract (`GainNode` unity gain, `EqualizerNode` with no bands, `ResampleNode` at a matching rate, `SpeedNode` unity speed).

`DynamicsNode.Process` no longer wraps the input in `using`. It tracks ownership and transfers it when a hook hands the buffer back (`src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs:100`), with a `finally` so the throwing paths and the empty-chunk early return still dispose it. Behaviour for the current derived nodes is unchanged — both always emit a fresh buffer, so `ReferenceEquals` is false and the input is disposed exactly as before.

Covered by `DynamicsNode_PassThroughHook_DoesNotDisposeReturnedInput`, which derives a node whose hooks return the input and asserts the caller receives a live buffer; it fails if ownership is unconditionally kept. Added `GateNode_DisposesConsumedInput` alongside it so the new node is pinned in that suite too.

File: src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+            }
+
+            // Nodes are cached across Compose() calls, so stale DSP state must not survive a seek or
+            // restart. Diagnostics latches are deliberately kept — re-arming them on every scrub would let
+            // a persistent fault re-log once per Process call.
+            if (!_lastTimeRangeEnd.HasValue || !context.ContinuesFrom(_lastTimeRangeEnd.Value))
+            {
+                ResetDspState();
+            }
+            _lastTimeRangeEnd = context.TimeRange.Start + context.TimeRange.Duration;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs]: Fixed in f06938f9f. `_lastTimeRangeEnd` is now assigned after the hook returns (`src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs:96`), so a chunk whose hook threw leaves the previous chunk's end recorded and the next call takes the discontinuity branch — matching the "Update only on success" comment on `LimiterNode.cs:123`.

Covered by the new `DynamicsNodeTests`, which derives a node whose hook can throw and whose `ResetDspState` counts calls: `Process_HookThrows_LeavesTheNextChunkNonContiguous` fails against the pre-fix ordering, and `Process_HookSucceeds_KeepsTheNextChunkContiguous` guards the other direction so the fix does not turn every chunk into a reset. Both count resets as a delta from the opening chunk, since that one resets for two independent reasons (sample-rate and no-predecessor) and the test should not depend on it.

File: src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+            {
+                ResetDspState();
+            }
+
+            AudioBuffer output = HasAnimatedParameters
+                ? ProcessAnimated(input, context)
+                : ProcessStatic(input, context);
+
+            // A chunk that threw must not look contiguous, or the next one inherits half-mutated state.
+            _lastTimeRangeEnd = context.TimeRange.Start + context.TimeRange.Duration;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/Graph/Nodes/DynamicsNode.cs:87]: Fixed in 9cea72744. DynamicsNode now clears `_lastTimeRangeEnd` whenever processing throws, so retrying the same timeline chunk is treated as non-contiguous and calls `ResetDspState()` before the hook runs again. Added `Process_HookThrows_RetryingSameChunkResetsDspState`, which failed against the previous code (expected reset count 3, observed 2) and now passes.
